Image tags

erikperkins opened the pull request:
Fixes issue #10.

Add tags to images and display tags along with the images.

shfung commented on app/views/images/new.html.erb:
```diff
@@ -1,5 +1,6 @@
 <%= form_for @image do |f| %>
   <%= f.label :url %>
   <%= f.text_field :url %>
+  <%= f.text_field :tag_list %>
```
Do we need label for this field?

erikperkins replied:
Addressed in 1e3fb9f7996ab7f3fc7aabe7243ade0d69f05525

shfung commented on config/initializers/acts_as_taggable_patch.rb:
```diff
@@ -0,0 +1 @@
+ActsAsTaggableOn.force_binary_collation = true
```
I'm wondering why do we need this setting?

erikperkins replied:
It's the suggested workaround https://github.com/mbleigh/acts-as-taggable-on#for-mysql-users for a bug in acts_as_taggable_on https://github.com/mbleigh/acts-as-taggable-on/issues/623. Heroku doesn't use MySQL, of course, but I included it anyway to be safe.

shfung commented on db/migrate/20160223173235_acts_as_taggable_on_migration.acts_as_taggable_on_engine.rb:
```diff
@@ -0,0 +1,31 @@
+class ActsAsTaggableOnMigration < ActiveRecord::Migration
+  def self.up
+    create_table :tags do |t|
+      t.string :name
```
I understand that these migration and table structures are provided by the gem, but I'm wondering wether we need to add some not-null constraints to some of the columns?

bboe replied:
> I'm wondering wether we need to add some not-null constraints to some of the columns

:+1:

erikperkins replied:
Addressed in dc45ed376e37a0667cbd53cb452ba06d429bd005

shfung commented on db/migrate/20160223173239_change_collation_for_tag_names.acts_as_taggable_on_engine.rb:
```diff
@@ -0,0 +1,10 @@
+# work properly
+class ChangeCollationForTagNames < ActiveRecord::Migration
+  def up
+    if ActsAsTaggableOn::Utils.using_mysql?
```
Do we need this migration?

erikperkins replied:
Given the current deployment environment, it's not necessary, but it's included for the same reason as the line in config/initializers/acts_as_taggable_patch.rb.

shfung commented on test/controllers/images_controller_test.rb:
```diff
@@ -3,17 +3,22 @@
+
+  test 'create image without tags' do
+    post :create, { image: { url: '', tag_list: '' } }
+    assert_response :unprocessable_entity
```
It seems that the content of this test doesn't match entirely with its name. The `unprocessable_entity` seems to be triggered by the empty url instead of the empty tag_lists.
Maybe we can add one more test with valid url but empty tag_list as params.

bboe replied:
I think it's okay if we have only one invalid test in the controller test since we rely on testing the effectiveness of each validator in the model test.

On a separate note, in such a failure, what else should we test for?

erikperkins replied:
Addressed in 03251743579f92c783e2f3aa43e0bb8b5e98fe40

I'm not sure what else we should test for, since the specification doesn't require images to have tags yet. The only way an image can be invalid (in this branch) is if it is submitted with a blank URL.

shfung commented on app/controllers/images_controller.rb:
```diff
@@ -6,16 +6,23 @@ def new
+    image = image_params
 
     @image = Image.new url: image[:url]
+    @image.tag_list.add image[:tag_list], parse: true
```
It seems that the `acts-as-taggable-on` gem allow any string to be the tag content.  Maybe we can consider sanitizing the string to avoid any potential SQL injection or XSS.

bboe replied:
@shfung I'm not sure if I agree that sanitizing things going into our database is the right approach. Sanitizing on display makes things much less complicated.

We should also be able to rely on `acts_as_taggable_on` to properly handle SQL injection. Of course we should verify that.

shfung replied:
Fair enough.

On a similar note: is there any format restriction or length limitation that we want to enforce on the tags?

bboe replied:
Each tag is a string in the rails migration. So whatever the database limit is on that field. I don't believe we explicitly test for field boundaries unless we have a reason to.

 Each image can have an any number of tags. A image with tens+ tags could be problematic, but that's not really something we're concerned with at this point (MVP).

erikperkins replied:
I'll leave this one alone for now.

rmacklin commented on test/models/image_test.rb:
```diff
@@ -12,4 +12,10 @@ class ImageTest < ActiveSupport::TestCase
+    url = 'http://imgs.xkcd.com/comics/haskell.png'
+    tags = ['haskell', 'programming']
+    assert_predicate Image.new(url: url, tag_list: tags), :valid?
+  end
```
I'm not sure of the value of this test as it's currently written.

erikperkins replied:
Addressed in 03251743579f92c783e2f3aa43e0bb8b5e98fe40

erikperkins commented:
@shfung @rmacklin @bboe How does it look now?

rmacklin commented:
> How does it look now?

It looks like appropriate squashing has not been done yet. Once that's done I can take another look.

erikperkins commented:
@rmacklin  Well, rebasing wiped out a bunch of your inline comments and my replies, but I rearranged things into more appropriate commits.

rmacklin commented:
@erikperkins Looks really good. I see that [these tests](https://github.com/AppFolioOnboarding/perkins_eidolon/blob/9a1a42ec398fe196b4998e500e40060ea884e5dc/test/controllers/images_controller_test.rb#L52-L60) made it back even though it had been mentioned that it could be overtesting to test the different validation errors at the controller level. But at this point, it's not a big deal so if you wanna leave it I don't mind.

erikperkins commented:
@rmacklin Okay, I changed the name of the test in [https://github.com/AppFolioOnboarding/perkins_eidolon/blob/7420a969090f665ff35e2025a7d9714201b3ee1a/test/models/image_test.rb, image_test.rb](url) and removed the two spurious tests in [https://github.com/AppFolioOnboarding/perkins_eidolon/blob/9a1a42ec398fe196b4998e500e40060ea884e5dc/test/controllers/images_controller_test.rb#L52-L60, image_controller_test.rb](url)

rmacklin commented:
:+1:

bboe commented on db/schema.rb:
```diff
@@ -19,4 +19,24 @@
     t.datetime "updated_at", null: false
   end
 
+  create_table "taggings", force: :cascade do |t|
```
Which of these fields can be set `null: False`? Ideally we'd like only fields that need `null: true` to allow null values.
